### ISSUE TITLE
Cannot do prediction using Torch Tensors

### ISSUE THREAD

**Issue by HCA97** [ultralytics/ultralytics#1226]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection

### Bug

First of all, thank you for your hard work.

I would like to run inference using `Torch Tensors`. According to the documentation, `Torch Tensor` is a supported source type (https://docs.ultralytics.com/predict/#sources). 

However, I got the following error:
```
Exception: Unsupported type encountered! See docs for supported types https://docs.ultralytics.com/predict
```
Am I doing something wrong, if so what might be the issue? 

Thank you for your time.

### Environment

- Ultralytics YOLOv8.0.48 🚀 Python-3.8.16 torch-1.13.1+cu117 CUDA:0 
- OS: Ubuntu 22.04

### Minimal Reproducible Example

```py
from ultralytics import YOLO
import torch as th


model = YOLO("yolov8n.pt")
image = th.rand((16,3,320,640))
results = model(image)
```

### Additional

I tried to access the `Torch Model` itself. However, when I get predictions I don't understand what the output is. 

How can I access the bounding boxes and the logits? 

```py
model = YOLO("yolov8n.pt").model.to('cuda')
image = th.rand((16,3,320,640), device='cuda')
results = model(image)
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@HCA97 thanks for the bug report, I'm able to reproduce using your code. 

There's no quick fix for this, as Tensor support is missing currently. In the meantime you can try a different source, i.e. a numpy array is probably closest to a torch tensor. The input formats supported are here:

https://docs.ultralytics.com/predict/


| source     | stream  | model(arg)                                 | type           | notes            |
|------------|---------|--------------------------------------------|----------------|------------------|
| image      |         | `'im.jpg'`                                 | `str`, `Path`  |                  |
| URL        |         | `'https://ultralytics.com/images/bus.jpg'` | `str`          |                  |
| screenshot |         | `'screen'`                                 | `str`          |                  |
| PIL        |         | `Image.open('im.jpg')`                     | `PIL.Image`    | HWC, RGB         |
| OpenCV     |         | `cv2.imread('im.jpg')[:,:,::-1]`           | `np.ndarray`   | HWC, BGR to RGB  |
| numpy      |         | `np.zeros((640,1280,3))`                   | `np.ndarray`   | HWC              |
| torch      |         | `torch.zeros(16,3,320,640)`                | `torch.Tensor` | BCHW, RGB        |
| CSV        |         | `'sources.csv'`                            | `str`, `Path`  | RTSP, RTMP, HTTP |         
| video      | &check; | `'vid.mp4'`                                | `str`, `Path`  |                  |
| directory  | &check; | `'path/'`                                  | `str`, `Path`  |                  |
| glob       | &check; | `path/*.jpg'`                              | `str`          | Use `*` operator |
| YouTube    | &check; | `'https://youtu.be/Zgi9g1ksQHc'`           | `str`          |                  |
| stream     | &check; | `'rtsp://example.com/media.mp4'`           | `str`          | RTSP, RTMP, HTTP |


@AyushExel @Laughing-q seems like we're missing Tensor inference support even though it's in the docs. Would this be easiest to add to numpy and PIL loader?

**Comment by glenn-jocher** [comment]
Duplicate of #1218

**Comment by HCA97** [comment]
Thank you for your quick reply. If that is fine with you, I would like to help with the implementation. :smile: Is there a direction you guys can point me?

**Comment by AyushExel** [comment]
> @HCA97 thanks for the bug report, I'm able to reproduce using your code.
> 
> There's no quick fix for this, as Tensor support is missing currently. In the meantime you can try a different source, i.e. a numpy array is probably closest to a torch tensor. The input formats supported are here:
> 
> https://docs.ultralytics.com/predict/
> 
> source	stream	model(arg)	type	notes
> image		`'im.jpg'`	`str`, `Path`	
> URL		`'https://ultralytics.com/images/bus.jpg'`	`str`	
> screenshot		`'screen'`	`str`	
> PIL		`Image.open('im.jpg')`	`PIL.Image`	HWC, RGB
> OpenCV		`cv2.imread('im.jpg')[:,:,::-1]`	`np.ndarray`	HWC, BGR to RGB
> numpy		`np.zeros((640,1280,3))`	`np.ndarray`	HWC
> torch		`torch.zeros(16,3,320,640)`	`torch.Tensor`	BCHW, RGB
> CSV		`'sources.csv'`	`str`, `Path`	RTSP, RTMP, HTTP
> video	✓	`'vid.mp4'`	`str`, `Path`	
> directory	✓	`'path/'`	`str`, `Path`	
> glob	✓	`path/*.jpg'`	`str`	Use `*` operator
> YouTube	✓	`'https://youtu.be/Zgi9g1ksQHc'`	`str`	
> stream	✓	`'rtsp://example.com/media.mp4'`	`str`	RTSP, RTMP, HTTP
> @AyushExel @Laughing-q seems like we're missing Tensor inference support even though it's in the docs. Would this be easiest to add to numpy and PIL loader?

Yeah I think we can add it. I'll attempt to

**Comment by AyushExel** [comment]
@HCA97 hey would something like this satisfy your use case https://github.com/ultralytics/ultralytics/pull/1252 ?

**Comment by glenn-jocher** [comment]
@AyushExel let's make an effort to link PRs to issues. They'll close automatically on PR merge also which is nice.

**Comment by HCA97** [comment]
Hi,

Thank you for your quick response. I have checked the unit test, and it is exactly what I need :)

I have a few questions about preprocessing. 
- Should I simply pass the tensors without normalization? 
- What size I should resize the images to? In the docs it says 320x640, can I use any size? 
- Do I need to keep the 1:2 aspect ratios?

**Comment by glenn-jocher** [comment]
@HCA97 with torch inputs the user is responsible for all preprocessing. There is no guidance we can provide other than to use the same preprocessing the steps the model saw during your training.

**Comment by glenn-jocher** [comment]
Removing TODO tag as issue is now fixed.

**Comment by HCA97** [comment]
Hi, I forgot to close the issue  😅  I can make predictions with torch tensors just the way I wanted. If you need to reopen the issue, feel free to do so.

**Comment by pderrenger** [comment]
@HCA97 no worries at all! Glad to hear that the issue is resolved and you can make predictions with torch tensors as desired. If you have any more questions or run into any other issues, feel free to open a new one. Thanks for reaching out!
